tf2_ros/include/tf2_ros/message_filter.h

```diff
@@ -491,6 +492,9 @@ class MessageFilter : public MessageFilterBase, public message_filters::SimpleFi
   {
     namespace mt = message_filters::message_traits;
 
+    // We will be accessing and mutating messages now, require unique lock
+    std::unique_lock<std::mutex> lock(messages_mutex_);
+
```

Comment:
As I pointed out in the comment, this isn't desirable since it holds onto the lock the entire time, even over `canTransform`.  However, if we look carefully at what this code does, it does looking for the correct item in the list, and assuming it finds it, it does an operation on it and removes it from the list.  So what I'll suggest we do instead here is something like the following pseudo-code:

```
{
    lock;
    for (; msg_it != msg_end; ++msg_it) {
        handle_it = std::find(info.handles.begin(), info.handles.end(), handle);
        if (handle_it != info.handles.end()) {
            info.success_count++;
            if (info.success_count >= expected_success_count_) {
              *saved_msg = msg_it;
              messages_.erase(msg_it);
              break;
            }
        }
    }
}
```

The idea here is to do all of the lookup and erasure of the item in the list under the lock at the beginning.  Then we drop the lock, and the rest of the code below operates on the thing we've removed from the list.  This should improve thread safety, but still not hold the lock across the `canTransform` calls.

Let me know if you think this will work, or don't understand what I am talking about here.

Reply:
Good point! I've made the adjustment, however this does require an additional copy of messages. ```saved_msg = *msg_it;```